File: src/main/java/at/xirado/bean/data/content/DismissableState.java
package at.xirado.bean.data.content;

public enum DismissableState {
    NONE(), // No state
    SEEN(), // The user has seen the content, but has not interacted with it.
    AWARE(), // The user is aware of the feature(s) showcased in this dismissable content.
    ACKNOWLEDGED() // The user explicitly acknowledged this dismissable content.
}

File: src/main/java/at/xirado/bean/data/content/DismissableState.kt
package at.xirado.bean.data.content

import at.xirado.bean.data.database.SQLBuilder
import java.sql.SQLException

class DismissableState(val userId: Long, val content: IDismissable<*>, status: Status) {
    var status: Status = status
        private set

    fun update(): DismissableState {
        try {
            SQLBuilder(UPDATE_QUERY, userId, content.identifier, status.name).execute()
        } catch (e: SQLException) {
            throw RuntimeException(e)
        }
        return this
    }

    fun updateStatus(status: Status): DismissableState {
        this.status = status
        return this.update()
    }

    companion object {
        private const val UPDATE_QUERY =
            "INSERT INTO dismissable_contents (user_id, identifier, state) values (?,?,?) " +
                    "ON DUPLICATE KEY UPDATE identifier = VALUES(identifier), state = VALUES(state) "
    }
}